server/aws-lsp-codewhisperer/src/language-server/agenticChat/tools/chatDb/chatDb.ts

```diff
@@ -510,7 +518,62 @@ export class ChatDatabase {
                 }
             }
         }
-        this.#features.logging.debug(`Current history characters: ${count}`)
+        return count
+    }
+
+    private calculateCurrentMessageCharacterCount(message: Message): number {
+        let count = 0
+        // Count characters of message text
+        count += message.body.length
+
+        // Count characters in tool uses
+        if (message.toolUses) {
+            try {
+                for (const toolUse of message.toolUses) {
+                    count += JSON.stringify(toolUse).length
+                }
+            } catch (e) {
+                this.#features.logging.error(`Error counting toolUses: ${String(e)}`)
+            }
+        }
+        // Count characters in tool results
+        if (message.userInputMessageContext?.toolResults) {
+            try {
+                for (const toolResul of message.userInputMessageContext.toolResults) {
+                    count += JSON.stringify(toolResul).length
+                }
+            } catch (e) {
+                this.#features.logging.error(`Error counting toolResults: ${String(e)}`)
+            }
+        }
+        // Count characters in tool spec for the current user message
+        if (message.userInputMessageContext?.tools) {
+            try {
+                for (const toolSpec of message.userInputMessageContext.tools) {
+                    count += JSON.stringify(toolSpec).length
+                }
+            } catch (e) {
+                this.#features.logging.error(`Error counting tool spec length: ${String(e)}`)
+            }
+        }
+
+        if (message.userInputMessageContext?.additionalContext) {
+            try {
+                for (const addtionalContext of message.userInputMessageContext.additionalContext) {
+                    count += JSON.stringify(addtionalContext).length
+                }
+            } catch (e) {
+                this.#features.logging.error(`Error counting addtionalContext length: ${String(e)}`)
+            }
+        }
+
+        if (message.userInputMessageContext?.editorState) {
```

Comment:
missing relevantDocuments (FYI, we will migrate to use that for falcon and deprecate additionalContext)